src/covjson_pydantic/ndarray.py

```diff
@@ -19,39 +18,35 @@ class DataType(str, Enum):
 class NdArray(CovJsonBaseModel):
     type: Literal["NdArray"] = "NdArray"
     dataType: DataType = DataType.float  # noqa: N815
-    axisNames: Optional[List[str]]  # noqa: N815
-    shape: Optional[List[int]]
+    axisNames: Optional[List[str]] = None  # noqa: N815
+    shape: Optional[List[int]] = None
     values: List[Optional[float]]
 
-    @root_validator(skip_on_failure=True)
-    def check_field_dependencies(cls, values):
-        if len(values["values"]) > 1 and (values.get("axisNames") is None or len(values.get("axisNames")) == 0):
+    @model_validator(mode="after")
+    def check_field_dependencies(self):
+        if len(self.values) > 1 and (self.axisNames is None or len(self.axisNames) == 0):
             raise ValueError("'axisNames' must to be provided if array is not 0D")
 
-        if len(values["values"]) > 1 and (values.get("shape") is None or len(values.get("shape")) == 0):
+        if len(self.values) > 1 and (self.shape is None or len(self.shape) == 0):
             raise ValueError("'shape' must to be provided if array is not 0D")
 
-        if (
-            values.get("axisNames") is not None
-            and values.get("shape") is not None
-            and len(values.get("axisNames")) != len(values.get("shape"))
-        ):
+        if self.axisNames is not None and self.shape is not None and len(self.axisNames) != len(self.shape):
             raise ValueError("'axisNames' and 'shape' should have equal length")
 
-        if values.get("shape") is not None and len(values.get("shape")) >= 1:
-            prod = math.prod(values["shape"])
-            if len(values["values"]) != prod:
+        if self.shape is not None and len(self.shape) >= 1:
+            prod = math.prod(self.shape)
+            if len(self.values) != prod:
                 raise ValueError(
                     "Where 'shape' is present and non-empty, the product of its values MUST equal "
                     "the number of elements in the 'values' array."
                 )
 
-        return values
+        return self
 
 
 class TileSet(BaseModel):
     tileShape: List[Optional[int]]  # noqa: N815
-    urlTemplate: AnyUrl  # noqa: N815
+    urlTemplate: str  # noqa: N815
```

Comment:
Is this change needed?

Reply:
Yes. See example in `spec-tiled-ndarray.json`. 
```
"urlTemplate": "http://example.com/c/{y}-{x}.covjson"
```
When you use AnyUrl. You get the following error:
```
E       assert '{"type":"Til...D.covjson"}]}' == '{"type":"Til...}.covjson"}]}'
E         Skipping 230 identical leading characters in diff, use -v to show
E         - ple.com/b/{t}.covjson"},{"tileShape":[null,2,3],"urlTemplate":"http://example.com/c/{y}-{x}.covjson"}]}
E         ?           ^ ^                                                                       ^ ^ ^ ^
E         + ple.com/b/%7Bt%7D.covjson"},{"tileShape":[null,2,3],"urlTemplate":"http://example.com/c/%7By%7D-%7Bx%7D.covjson"}]}
E         ?           ^^^ ^^^                                                                       ^^^ ^^^ ^^^ ^^^
```
So, it's URL encoding in between. But we don't want this here, because it's technically not an URL yet. It's a template.